// @connext/cf-core/dist/src/protocol/setup.js
"use strict";
var __await = (this && this.__await) || function (v) { return this instanceof __await ? (this.v = v, this) : new __await(v); }
var __asyncGenerator = (this && this.__asyncGenerator) || function (thisArg, _arguments, generator) {
    if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
    var g = generator.apply(thisArg, _arguments || []), i, q = [];
    return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
    function verb(n) { if (g[n]) i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
    function resume(n, v) { try { step(g[n](v)); } catch (e) { settle(q[0][3], e); } }
    function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
    function fulfill(value) { resume("next", value); }
    function reject(value) { resume("throw", value); }
    function settle(f, v) { if (f(v), q.shift(), q.length) resume(q[0][0], q[0][1]); }
};
Object.defineProperty(exports, "__esModule", { value: true });
const ethereum_1 = require("../ethereum");
const machine_1 = require("../machine");
const enums_1 = require("../machine/enums");
const state_channel_1 = require("../models/state-channel");
const signature_forwarder_1 = require("./utils/signature-forwarder");
const signature_validator_1 = require("./utils/signature-validator");
const protocol = enums_1.Protocol.Setup;
const { OP_SIGN, IO_SEND, IO_SEND_AND_WAIT, PERSIST_STATE_CHANNEL } = enums_1.Opcode;
exports.SETUP_PROTOCOL = {
    0: function (context) {
        return __asyncGenerator(this, arguments, function* () {
            const { message, network } = context;
            const { processID, params } = message;
            const { multisigAddress, responderXpub, initiatorXpub } = params;
            const stateChannel = state_channel_1.StateChannel.setupChannel(network.IdentityApp, multisigAddress, [initiatorXpub, responderXpub]);
            const setupCommitment = new ethereum_1.SetupCommitment(network, stateChannel.multisigAddress, stateChannel.multisigOwners, stateChannel.freeBalance.identity);
            const initiatorSignature = yield yield __await([OP_SIGN, setupCommitment]);
            const { customData: { signature: responderSignature } } = yield yield __await([
                IO_SEND_AND_WAIT,
                {
                    protocol,
                    processID,
                    params,
                    seq: 1,
                    toXpub: responderXpub,
                    customData: {
                        signature: initiatorSignature
                    }
                }
            ]);
            signature_validator_1.assertIsValidSignature(machine_1.xkeyKthAddress(responderXpub, 0), setupCommitment, responderSignature);
            yield yield __await([PERSIST_STATE_CHANNEL, [stateChannel]]);
            context.stateChannelsMap.set(stateChannel.multisigAddress, stateChannel);
        });
    },
    1: function (context) {
        return __asyncGenerator(this, arguments, function* () {
            const { message, network } = context;
            const { processID, params, customData: { signature: initiatorSignature } } = message;
            const { multisigAddress, initiatorXpub, responderXpub } = params;
            const stateChannel = state_channel_1.StateChannel.setupChannel(network.IdentityApp, multisigAddress, [initiatorXpub, responderXpub]);
            const setupCommitment = new ethereum_1.SetupCommitment(network, stateChannel.multisigAddress, stateChannel.multisigOwners, stateChannel.freeBalance.identity);
            signature_validator_1.assertIsValidSignature(machine_1.xkeyKthAddress(initiatorXpub, 0), setupCommitment, initiatorSignature);
            const responderSignature = yield yield __await([OP_SIGN, setupCommitment]);
            yield yield __await([PERSIST_STATE_CHANNEL, [stateChannel]]);
            yield yield __await([
                IO_SEND,
                {
                    protocol,
                    processID,
                    toXpub: initiatorXpub,
                    seq: signature_forwarder_1.UNASSIGNED_SEQ_NO,
                    customData: {
                        signature: responderSignature
                    }
                }
            ]);
            context.stateChannelsMap.set(stateChannel.multisigAddress, stateChannel);
        });
    }
};
