// @connext/cf-core/dist/src/protocol/update.js
"use strict";
var __await = (this && this.__await) || function (v) { return this instanceof __await ? (this.v = v, this) : new __await(v); }
var __asyncGenerator = (this && this.__asyncGenerator) || function (thisArg, _arguments, generator) {
    if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
    var g = generator.apply(thisArg, _arguments || []), i, q = [];
    return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
    function verb(n) { if (g[n]) i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
    function resume(n, v) { try { step(g[n](v)); } catch (e) { settle(q[0][3], e); } }
    function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
    function fulfill(value) { resume("next", value); }
    function reject(value) { resume("throw", value); }
    function settle(f, v) { if (f(v), q.shift(), q.length) resume(q[0][0], q[0][1]); }
};
Object.defineProperty(exports, "__esModule", { value: true });
const ethereum_1 = require("../ethereum");
const machine_1 = require("../machine");
const enums_1 = require("../machine/enums");
const signature_forwarder_1 = require("./utils/signature-forwarder");
const signature_validator_1 = require("./utils/signature-validator");
const protocol = enums_1.Protocol.Update;
const { OP_SIGN, IO_SEND, IO_SEND_AND_WAIT, PERSIST_STATE_CHANNEL } = enums_1.Opcode;
exports.UPDATE_PROTOCOL = {
    0: function (context) {
        return __asyncGenerator(this, arguments, function* () {
            const { stateChannelsMap, message, network } = context;
            const { processID, params } = message;
            const { appIdentityHash, multisigAddress, responderXpub, newState } = params;
            const preProtocolStateChannel = stateChannelsMap.get(multisigAddress);
            const postProtocolStateChannel = preProtocolStateChannel.setState(appIdentityHash, newState);
            const appInstance = postProtocolStateChannel.getAppInstance(appIdentityHash);
            const setStateCommitment = new ethereum_1.SetStateCommitment(network, appInstance.identity, appInstance.hashOfLatestState, appInstance.versionNumber, appInstance.timeout);
            const initiatorSignature = yield yield __await([
                OP_SIGN,
                setStateCommitment,
                appInstance.appSeqNo
            ]);
            const { customData: { signature: responderSignature } } = yield yield __await([
                IO_SEND_AND_WAIT,
                {
                    protocol,
                    processID,
                    params,
                    seq: 1,
                    toXpub: responderXpub,
                    customData: {
                        signature: initiatorSignature
                    }
                }
            ]);
            signature_validator_1.assertIsValidSignature(machine_1.xkeyKthAddress(responderXpub, appInstance.appSeqNo), setStateCommitment, responderSignature);
            yield yield __await([PERSIST_STATE_CHANNEL, [postProtocolStateChannel]]);
            context.stateChannelsMap.set(postProtocolStateChannel.multisigAddress, postProtocolStateChannel);
        });
    },
    1: function (context) {
        return __asyncGenerator(this, arguments, function* () {
            const { stateChannelsMap, message, network } = context;
            const { processID, params, customData: { signature: initiatorSignature } } = message;
            const { appIdentityHash, multisigAddress, initiatorXpub, newState } = params;
            const preProtocolStateChannel = stateChannelsMap.get(multisigAddress);
            const postProtocolStateChannel = preProtocolStateChannel.setState(appIdentityHash, newState);
            const appInstance = postProtocolStateChannel.getAppInstance(appIdentityHash);
            const setStateCommitment = new ethereum_1.SetStateCommitment(network, appInstance.identity, appInstance.hashOfLatestState, appInstance.versionNumber, appInstance.timeout);
            signature_validator_1.assertIsValidSignature(machine_1.xkeyKthAddress(initiatorXpub, appInstance.appSeqNo), setStateCommitment, initiatorSignature);
            const responderSignature = yield yield __await([
                OP_SIGN,
                setStateCommitment,
                appInstance.appSeqNo
            ]);
            yield yield __await([PERSIST_STATE_CHANNEL, [postProtocolStateChannel]]);
            yield yield __await([
                IO_SEND,
                {
                    protocol,
                    processID,
                    toXpub: initiatorXpub,
                    seq: signature_forwarder_1.UNASSIGNED_SEQ_NO,
                    customData: {
                        signature: responderSignature
                    }
                }
            ]);
            context.stateChannelsMap.set(postProtocolStateChannel.multisigAddress, postProtocolStateChannel);
        });
    }
};
