// @connext/cf-core/dist/test/machine/integration/random-signing-keys.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const utils_1 = require("ethers/utils");
const hdnode_1 = require("ethers/utils/hdnode");
const xkeys_1 = require("../../../src/machine/xkeys");
function getSortedRandomSigningKeys(length) {
    return Array(length)
        .fill(0)
        .map(_ => new utils_1.SigningKey(utils_1.hexlify(utils_1.randomBytes(32))))
        .sort((a, b) => parseInt(a.address, 16) < parseInt(b.address, 16) ? -1 : 1);
}
exports.getSortedRandomSigningKeys = getSortedRandomSigningKeys;
function extendedPrvKeyToExtendedPubKey(extendedPrvKey) {
    return hdnode_1.fromExtendedKey(extendedPrvKey).neuter().extendedKey;
}
exports.extendedPrvKeyToExtendedPubKey = extendedPrvKeyToExtendedPubKey;
function getRandomExtendedPubKey() {
    return extendedPrvKeyToExtendedPubKey(xkeys_1.computeRandomExtendedPrvKey());
}
exports.getRandomExtendedPubKey = getRandomExtendedPubKey;
function getRandomExtendedPubKeys(length) {
    return Array(length)
        .fill(0)
        .map(getRandomExtendedPubKey);
}
exports.getRandomExtendedPubKeys = getRandomExtendedPubKeys;
function getRandomExtendedPrvKeys(length) {
    return Array(length)
        .fill(0)
        .map(xkeys_1.computeRandomExtendedPrvKey);
}
exports.getRandomExtendedPrvKeys = getRandomExtendedPrvKeys;
function getRandomHDNodes(length) {
    return getRandomExtendedPrvKeys(length).map(x => hdnode_1.fromExtendedKey(x));
}
exports.getRandomHDNodes = getRandomHDNodes;
