// @connext/cf-core/dist/test/integration/unidirectional-transfer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const constants_1 = require("ethers/constants");
const utils_1 = require("ethers/utils");
const singleAssetTwoPartyCoinTransferEncoding = `
tuple(address to, uint256 amount)[2]
`;
exports.transferAbiEncodings = {
    stateEncoding: `
    tuple(
      uint8 stage,
      ${singleAssetTwoPartyCoinTransferEncoding} transfers,
      uint256 turnNum,
      bool finalized
    )`,
    actionEncoding: `
    tuple(
      uint8 actionType,
      uint256 amount
    )`
};
function validAction(actionType = 0, amount = 1) {
    return {
        actionType,
        amount
    };
}
exports.validAction = validAction;
function initialTransferState(senderAddr, receiverAddr, amount = 1) {
    return {
        stage: 0,
        transfers: [
            {
                amount: utils_1.bigNumberify(amount),
                to: senderAddr
            },
            {
                to: receiverAddr,
                amount: constants_1.Zero
            }
        ],
        turnNum: 0,
        finalized: false
    };
}
exports.initialTransferState = initialTransferState;
