// @connext/types/dist/src/app.js
export const SupportedApplications = {
    CoinBalanceRefundApp: "CoinBalanceRefundApp",
    SimpleLinkedTransferApp: "SimpleLinkedTransferApp",
    SimpleTransferApp: "SimpleTransferApp",
    SimpleTwoPartySwapApp: "SimpleTwoPartySwapApp",
};
export const SupportedNetworks = {
    ganache: "ganache",
    goerli: "goerli",
    homestead: "homestead",
    kovan: "kovan",
    rinkeby: "rinkeby",
    ropsten: "ropsten",
};
export var UnidirectionalTransferAppActionType;
(function (UnidirectionalTransferAppActionType) {
    UnidirectionalTransferAppActionType[UnidirectionalTransferAppActionType["SEND_MONEY"] = 0] = "SEND_MONEY";
    UnidirectionalTransferAppActionType[UnidirectionalTransferAppActionType["END_CHANNEL"] = 1] = "END_CHANNEL";
})(UnidirectionalTransferAppActionType || (UnidirectionalTransferAppActionType = {}));
export var UnidirectionalTransferAppStage;
(function (UnidirectionalTransferAppStage) {
    UnidirectionalTransferAppStage[UnidirectionalTransferAppStage["POST_FUND"] = 0] = "POST_FUND";
    UnidirectionalTransferAppStage[UnidirectionalTransferAppStage["MONEY_SENT"] = 1] = "MONEY_SENT";
    UnidirectionalTransferAppStage[UnidirectionalTransferAppStage["CHANNEL_CLOSED"] = 2] = "CHANNEL_CLOSED";
})(UnidirectionalTransferAppStage || (UnidirectionalTransferAppStage = {}));
export var UnidirectionalLinkedTransferAppStage;
(function (UnidirectionalLinkedTransferAppStage) {
    UnidirectionalLinkedTransferAppStage[UnidirectionalLinkedTransferAppStage["POST_FUND"] = 0] = "POST_FUND";
    UnidirectionalLinkedTransferAppStage[UnidirectionalLinkedTransferAppStage["PAYMENT_CLAIMED"] = 1] = "PAYMENT_CLAIMED";
    UnidirectionalLinkedTransferAppStage[UnidirectionalLinkedTransferAppStage["CHANNEL_CLOSED"] = 2] = "CHANNEL_CLOSED";
})(UnidirectionalLinkedTransferAppStage || (UnidirectionalLinkedTransferAppStage = {}));
