// @connext/cf-core/dist/src/methods/app-instance/reject-install/controller.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const rpc_server_1 = require("rpc-server");
const request_handler_1 = require("../../../request-handler");
const types_1 = require("../../../types");
const controller_1 = require("../../controller");
const controller_2 = __importDefault(require("../reject-install-virtual/controller"));
class RejectInstallController extends controller_1.NodeController {
    async getRequiredLockNames(requestHandler, params) {
        const { appInstanceId } = params;
        return [appInstanceId];
    }
    async executeMethodImplementation(requestHandler, params) {
        const { store, messagingService, publicIdentifier } = requestHandler;
        const { appInstanceId } = params;
        const appInstanceProposal = await store.getAppInstanceProposal(appInstanceId);
        if (appInstanceProposal.intermediaryIdentifier) {
            return controller_2.default(requestHandler, params);
        }
        const stateChannel = await store.getChannelFromAppInstanceID(appInstanceId);
        await store.saveStateChannel(stateChannel.removeProposal(appInstanceId));
        const rejectProposalMsg = {
            from: publicIdentifier,
            type: types_1.NODE_EVENTS.REJECT_INSTALL,
            data: {
                appInstanceId
            }
        };
        await messagingService.send(appInstanceProposal.proposedByIdentifier, rejectProposalMsg);
        return {};
    }
}
__decorate([
    rpc_server_1.jsonRpcMethod(types_1.Node.RpcMethodName.REJECT_INSTALL),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [request_handler_1.RequestHandler, Object]),
    __metadata("design:returntype", Promise)
], RejectInstallController.prototype, "executeMethodImplementation", null);
exports.default = RejectInstallController;
