// @connext/cf-core/dist/src/ethereum/withdraw-erc20-commitment.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const utils_1 = require("ethers/utils");
const contracts_1 = require("../contracts");
const multisig_commitment_1 = require("./multisig-commitment");
const types_1 = require("./types");
class WithdrawERC20Commitment extends multisig_commitment_1.MultisigCommitment {
    constructor(multisigAddress, multisigOwners, to, value, tokenAddress) {
        super(multisigAddress, multisigOwners);
        this.multisigAddress = multisigAddress;
        this.multisigOwners = multisigOwners;
        this.to = to;
        this.value = value;
        this.tokenAddress = tokenAddress;
    }
    getTransactionDetails() {
        return {
            data: new utils_1.Interface(contracts_1.ERC20.abi).functions.transfer.encode([this.to, this.value]),
            operation: types_1.MultisigOperation.Call,
            to: this.tokenAddress,
            value: 0,
        };
    }
}
exports.WithdrawERC20Commitment = WithdrawERC20Commitment;
